### DISCUSSION TITLE
How  to run tests on electron desktop app with wdio

### DISCUSSION THREAD

**Discussion by ellatom** [webdriverio/webdriverio#8058] [Technical Discussion]
Hi, I need help. I'm trying to run tests on electron desktop app with wdio, however i'm getting error:
```
Error: Failed to create session.
[0-0] unknown error: Chrome failed to start: exited normally.
[0-0]   (unknown error: DevToolsActivePort file doesn't exist)
```
What could be the reason?
Thanks.

**Comment by christian-bromann** [comment]
@ellatom are you using the `wdio-electron-service`? On what system are you running it?

**Comment by ellatom** [comment]
@christian-bromann  Yes. I have `"wdio-chromedriver-service": "^7.2.8",` , I'm running on windows os. 
I had few errors including the one above
After I have added to debug arg to
```
 'goog:chromeOptions': {
            args: [
            '--remote-debugging-port=9222'] 
```
 The errors stopped to appear.

**Reply by christian-bromann** [reply]
We recommend to use the [wdio-__electron__-service`](https://www.npmjs.com/package/wdio-electron-service) to run tests for Electron apps.

**Reply by ellatom** [reply]
@christian-bromann I'm building automation for electron desktop app. When I use the wdio-electron-service the opened url is empty, thus unable to simulate user dom actions.
Only when I use chromedriver service the desktop app is being opened . Do you have an example from github projects with this service that can open an 'exe' app and run on top of it automation? Even in the webdriver io documentation the project for desktop was with this driver [AdarshKumarGM/WebDriverIO-with-CucumberJS](https://github.com/AdarshKumarGM/WebDriverIO-with-CucumberJS)[​](https://webdriver.io/docs/boilerplates#adarshkumargmwebdriverio-with-cucumberjs) . Maybe I misunderstood something?

**Reply by christian-bromann** [reply]
> Maybe I misunderstood something?

Not sure why the opened url is empty but the service is build so you don't have to wrestle with all these configuration. This service was implemented recently and is our recommendation testing electron apps moving forward.

**Reply by ellatom** [reply]
@christian-bromann Do you have an example project to look on how to configure my project, so that I will be able to run on a windows desktop application(which opens via exe file) , some automation tests with 'wdio-electron-service' and not 'chromedriver' service?
This is my current wdio.conf.js:
```
exports.config = {
    //
    // ====================
    // Runner Configuration
    // ====================
    specs: [
        './test/specs/**/*.js'
    ],
    // Patterns to exclude.
    exclude: [
        // 'path/to/excluded/files'
    ],
    maxInstances: 1,

    capabilities: [{

        maxInstances: 1,
       
        browserName: 'chrome',
        'goog:chromeOptions': {
            binary: '../../AppData/Local/Ella/ella.exe',
            args: ['--disable-infobars'
            ,'--disable-dev-shm-usage'
            ,'--no-sandbox',
            '--remote-debugging-port=9222'] // Optional, perhaps 'app=' + /path/to/your/app/
          },
        acceptInsecureCerts: true
    }],
   
    bail: 0,
  
    waitforTimeout: 10000,

    connectionRetryTimeout: 120000,
  
    connectionRetryCount: 2,
    
    services: [['chromedriver', {
        chromedriverCustomPath:"./node_modules/electron-chromedriver/bin/chromedriver.exe",
        // path: "/wd/hub/",
        port:4444
    }]],
  
    mochaOpts: {
        ui: 'bdd',
        timeout: 60000
    }
}
```

**Reply by christian-bromann** [reply]
please check the example configuration in the [service docs](https://github.com/webdriverio-community/wdio-electron-service#example-configuration)

**Reply by ellatom** [reply]
@christian-bromann I have added the configuration from the example and made some adjustments. Currently the desktop electron is being opened, however only browser.getTitle() return the name of the title. Function like $(element selector).click() doesn't happen on the GUI. Visually I don't see that something as expected is happen after the click. It looks like the action isn't being simulated on the gui.
this is my config- maybe I haven't added something to the config- Please advise:
```
kill = require("kill-port");
const { join } = require('path');
// const fs = require('fs');

exports.config = {
    specs: [
        './test/specs/**/*.js'
    ],
    exclude: [
        // 'path/to/excluded/files'
    ],
    maxInstances: 1,
    capabilities: [{
        maxInstances: 1,
        acceptInsecureCerts: true
    }],
    logLevel: 'info',
    debug: true,
    bail: 0,
    waitforTimeout: 100000,
    connectionRetryTimeout: 120000,
    connectionRetryCount: 2,
    outputDir: 'all-logs',

    services: [['electron', {
        binaryPath:'../../AppData/Local/Ella/Ella.exe',
        appArgs: ['--disable-infobars'
                ,'--disable-dev-shm-usage'
                ,'--no-sandbox',
                '--remote-debugging-port=9222'],
        chromedriver: {
            port: 9519,
            logFileName: 'wdio-chromedriver.log',
            },       
    }]],
  port: 9519,
    framework: 'mocha',
    reporters: ['spec'],
    mochaOpts: {
        ui: 'bdd',
        timeout: 60000
    },
    onComplete: async function (exitCode, config, capabilities, results) {
        await kill(4444, 'tcp')
    },
}
```

**Reply by christian-bromann** [reply]
@goosewobbler do you see anything problematic here?

**Reply by goosewobbler** [reply]
Not really, I think this might be a bug actually - I just moved over the clicking interaction tests I had in the Spectron fork to my local copy of `wdio-electron-service-example` and they time out...

UPDATE:
Seems they're timing out because of reliance on the browserWindow API which I just ported (and is failing). Currently getting no logs through from anything wrapped in executeAsync so API progress is slow atm, but separating the clicking from the API stuff shows that the clicks are working. The clicking issue reported here could be due to a race condition between the test execution and render of the UI - I removed the spectron `waitUntilWindowLoaded` helper so maybe @ellatom can try putting a delay of a few seconds at the start of the test.

**Reply by ellatom** [reply]
@goosewobbler Thanks for your fast reply. I have added 'debugger' and wait until I see the app with the content inside, however the click doesn't happen inside the app. The test completes with success, but the click doesn't happen.

**Reply by ellatom** [reply]
@goosewobbler I have added few validation and it finally works.
I wanted to make sure that the element is clickable and existing(found) before I try to make a click.
The strange thing is that each time I use the runner with command 
`npx wdio run wdio.conf.js` , 3 windows of same app(in task manager it shows as 3 processes) opened and only one of the windows is actually do the simulated action(for example the click)- Is this the expected result? Is there a way to open/show only 1 window -the one on which the commands are being made?
```
const el = await $('ul>li:nth-child(2)>span[class^=styles_tab-title]')
        let clickable = await el.isClickable();
        console.log(clickable); // outputs: true or false
        let isExisting = await el.isExisting();
        console.log(isExisting);
        el.click()
```

**Reply by goosewobbler** [reply]
@ellatom this is expected - current behaviour reloads the browser session between tests to avoid state bleed. I might add a prop to change this back to the previous behaviour, in this case you will have to manage your own state and assume that every stateful action will have an effect on the next test.

**Reply by goosewobbler** [reply]
Raised https://github.com/webdriverio-community/wdio-electron-service/issues/21 to cover this.

**Reply by ellatom** [reply]
@goosewobbler I'm newbie in wdio, but why when I'm running on same session only one test, more than 1 window opens? Is there a way to prevent this?

**Reply by goosewobbler** [reply]
I believe WDIO opens a worker process for each spec file.  Each worker has a session which would normally be a single instance of the Electron app - so if you have four spec files you would get four instances.  The instances are closed once all the tests in a given spec file are complete.  The current behaviour of the Electron service reloads the session for a given spec file, between each test which spawns a new instance for each individual test and then closes it again.  The point of this is to avoid state bleed.

If you've only got a single spec file with a single test in it, you should only have one single instance of Electron running.

The proposal in https://github.com/webdriverio-community/wdio-electron-service/issues/21 is to allow for the reload behaviour to be configurable. Initially I think we keep the current behaviour and add a prop to switch off the session reloading.  This would mean if you want you can revert to the previous behaviour of one session per spec file.  Depending on what the community find preferable we can reverse the logic in future and it default to no session reloading.

**Reply by ellatom** [reply]
@goosewobbler I have 1 spec only, however more than 1 instance of the app created( meaning 4 app windows are being opened). 
I have added inside test this code to retrieve the list of all window handles available to the session and it looks like there is only 1:

```
var windowHandles = await browser.getWindowHandle()
console.log(windowHandles);
//prints  CDwindow-770B840682A809D6256FB5C999FAD66C
```

**Reply by goosewobbler** [reply]
@ellatom how many tests do you have in your spec file?

It makes sense that there would be only one window handle, because only one window per session is being created - however the session is being reloaded after each test which gives you a new window (and handle).

**Reply by ellatom** [reply]
@goosewobbler I have 1 spec with 1 test. Is there a way to stop from session to be reloaded? what is the key and value should I add to wdio.conf.js ? I can try to stop the reloading and see if this the issue.

**Reply by goosewobbler** [reply]
@ellatom ok so you shouldn't get more than one window then.  If you had four tests then that would make sense.  However without seeing all of your code I have no idea if there is something else causing it.

The session reload behaviour is not yet configurable, as I said I raised an issue for this, I can work on it after the API work is complete.

You can change the behaviour locally by going into your `node_modules/wdio-electron-service/dist` dir and find the `service.js` file - you need to comment out or delete this:

```
    async afterTest() {
        await browser?.reloadSession();
    }
```

**Reply by ellatom** [reply]
@goosewobbler I have commented out and 1 instance was removed. Currently only 3 instances created. It means that there are 2 more instances than expected. Can you help to understand why I still have 2 more than expected?

**Reply by goosewobbler** [reply]
@ellatom without seeing your code, no.  I have no idea why it might be happening.  Basically the example repo I set up is behaving as expected, so I'm thinking it's something to do with your setup or usage of WDIO.  If you can provide a minimal git repository which reproduces the problem then that would help a lot.

**Reply by ellatom** [reply]
@goosewobbler This is the project. I can't add the electron app, so in the configuration file under binary path key a path to electron app exe  file should be added. Let me know if you think that I need to change something/update in the project, so I won't get 3 instances of the same app.
[ElectronAutomation.zip](https://github.com/webdriverio/webdriverio/files/8176781/ElectronAutomation.zip)

**Reply by goosewobbler** [reply]
@ellatom Ok, just looking at this, you have `chromedriver` and `electron-chromedriver` as devdeps - should work without these and you shouldn't need to specify `chromedriverCustomPath` in the WDIO config, that's already done for you.  You also shouldn't need `wdio-chromedriver-service` as a dep because that's provided with the service.

EDIT:
Fixed an issue with `electron-chromedriver`.  Try again after installing the new version and removing reloadSession - now works for me with the above changes.

**Reply by ellatom** [reply]
@goosewobbler I have made the suggested adjustments above, However I'm still getting 3 instances of the app.
Just to make sure what is the electron-chromedriver version you are using?

**Reply by goosewobbler** [reply]
No electron-chromedriver. As I said above you shouldn't need it. I suspect you installed it because of the bug I just fixed actually.

```
  "dependencies": {
    "chai": "^4.3.6",
    "eslint": "^8.10.0",
    "fs": "^0.0.2",
    "kill-port": "^1.6.1",
    "node-modules": "^1.0.1",
    "path": "^0.12.7"
  },
  "devDependencies": {
    "@types/mocha": "^9.1.0",
    "@types/node": "^17.0.21",
    "@typescript-eslint/eslint-plugin": "^5.13.0",
    "@typescript-eslint/parser": "^5.13.0",
    "@wdio/cli": "^7.16.15",
    "@wdio/local-runner": "^7.16.15",
    "@wdio/mocha-framework": "^7.16.15",
    "@wdio/spec-reporter": "^7.16.14",
    "ts-node": "^10.6.0",
    "typescript": "^4.6.2",
    "wdio-electron-service": "^2.0.7",
    "wdio-selenium-standalone-service": "^0.0.12",
    "webdriverio": "^7.16.16"
  }
```

**Reply by ellatom** [reply]
@goosewobbler I have uninstalled and this is the current package.json:
```{
  "name": "electronautomation",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "npm run"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "chai": "^4.3.6",
    "electron": "^17.1.0",
    "eslint": "^8.10.0",
    "node-modules": "^1.0.1"
  },
  "devDependencies": {
    "@types/mocha": "^9.1.0",
    "@types/node": "^17.0.21",
    "@typescript-eslint/eslint-plugin": "^5.13.0",
    "@typescript-eslint/parser": "^5.13.0",
    "@wdio/cli": "^7.16.15",
    "@wdio/local-runner": "^7.16.15",
    "@wdio/mocha-framework": "^7.16.15",
    "@wdio/spec-reporter": "^7.16.14",
    "ts-node": "^10.6.0",
    "typescript": "^4.6.2",
    "wdio-electron-service": "^2.0.5",
    "webdriverio": "^7.16.15"
  }
}
```
and this is my wdio.conf.ts:
```
exports.config = {
    autoCompileOpts: {
        autoCompile: true,

        tsNodeOpts: {
            transpileOnly: true,
            project: 'tsconfig.json'
        },
    },
  
    specs: [
        './test/specs/**/*.ts'
    ],
    // Patterns to exclude.
    exclude: [
        // 'path/to/excluded/files'
    ],

    maxInstances: 1,

    capabilities: [{
    }],
    logLevel: 'error',
    debug: true,
  
    bail: 0,
  
    waitforTimeout: 200000,
    
    connectionRetryTimeout: 600000,
  
    connectionRetryCount: 2,
      
    outputDir: 'all-logs',

    services: [['electron', {
        binaryPath:'../../AppData/Local/Workstation/Workstation.exe',
       
        appArgs: [
                '--disable-infobars'
                ,'--disable-dev-shm-usage'
                ,'--no-sandbox',
                '--remote-debugging-port=9222'],
        chromedriver: {
            // chromedriverCustomPath:"./node_modules/electron-chromedriver/bin/chromedriver.exe",
            port: 9519,
            logFileName: 'wdio-chromedriver.log',
            },       

    }]],
  port: 9519,
   
    framework: 'mocha',

    reporters: ['spec'],

    mochaOpts: {
        ui: 'bdd',
        timeout: 60000
    },
  
     execArgv: ['--inspect'],
}
```

I'm still seeing the same issue.
I tried to debug and I see that in @wdio->runner->build->index.js after start session  3 instances of app are being opened.
```
 async _initSession(config, caps, browserStub) {
        debugger;
        const browser = await this._startSession(config, caps);
```

**Reply by goosewobbler** [reply]
@ellatom I'm sorry but this works on my machine. I've already gone way over and above what most package maintainers would do in helping you debug this. I'll put in the option for configuring the reloadSession so that you don't have to comment that out any more but I've got a lot of other work to be doing.

**Reply by goosewobbler** [reply]
This is the config I was using, node modules I already specified.

```
exports.config = {
  autoCompileOpts: {
    autoCompile: true,
    tsNodeOpts: {
      transpileOnly: true,
      project: "tsconfig.json",
    },
  },
  specs: ["./test/specs/**/*.ts"],
  exclude: [],
  maxInstances: 1,
  capabilities: [
    {
      acceptInsecureCerts: true,
    },
  ],
  logLevel: "info",
  debug: true,
  bail: 0,
  waitforTimeout: 200000,
  connectionRetryTimeout: 120000,
  connectionRetryCount: 2,
  outputDir: "all-logs",
  services: [
    [
      "electron",
      {
        binaryPath:
          "../../workspace/wdio-electron-service-example/dist/mac/wdio-electron-service-example.app/Contents/MacOS/wdio-electron-service-example",
        appArgs: [
          "--disable-infobars",
          "--disable-dev-shm-usage",
          "--no-sandbox",
          "--remote-debugging-port=9222",
        ],
        chromedriver: {
          port: 9521,
          logFileName: "wdio-chromedriver.log",
        },
      },
    ],
  ],
  port: 9521,
  framework: "mocha",
  reporters: ["spec"],
  mochaOpts: {
    ui: "bdd",
    timeout: 60000,
  },
};

```

**Reply by ellatom** [reply]
@goosewobbler Thanks for your help.

**Reply by goosewobbler** [reply]
@ellatom I did a quick release of the session reload pref.  If you upgrade to 2.1.0 you should be able to turn off the session reload by specifying `newSessionPerTest: false` in the service config.
https://github.com/webdriverio-community/wdio-electron-service#newsessionpertest-boolean
